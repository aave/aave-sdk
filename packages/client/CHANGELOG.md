# @aave/client

## 0.3.1

### Patch Changes

- **fix:** Support max and exact value options in repay function.
- Updated dependencies
  - @aave/graphql@0.3.1

## 0.3.0

### Minor Changes

- **feat:** support for healthFactorPreview query

### Patch Changes

- Updated dependencies
  - @aave/graphql@0.3.0

## 0.2.0

### Minor Changes

- 0e36ce5: **feat:** supports for borrow credit delegation

### Patch Changes

- Updated dependencies [0e36ce5]
  - @aave/graphql@0.2.0

## 0.1.0

### Minor Changes

- fb26904: **feat:** first release

### Patch Changes

- Updated dependencies [fb26904]
  - @aave/graphql@0.1.0
  - @aave/types@0.1.0
