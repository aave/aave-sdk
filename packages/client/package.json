{
  "name": "@aave/client",
  "version": "0.0.0",
  "description": "The official JavaScript client for the Aave API",
  "repository": {
    "directory": "packages/client",
    "type": "git",
    "url": "git://github.com/aave/aave-sdk.git"
  },
  "type": "module",
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    },
    "./actions": {
      "import": "./dist/actions/index.js",
      "require": "./dist/actions/index.cjs"
    },
    "./ethers": {
      "import": "./dist/ethers.js",
      "require": "./dist/ethers.cjs"
    },
    "./viem": {
      "import": "./dist/viem.js",
      "require": "./dist/viem.cjs"
    },
    "./test-utils": {
      "import": "./dist/test-utils.js",
      "require": "./dist/test-utils.cjs"
    }
  },
  "typesVersions": {
    "*": {
      "actions": [
        "./dist/actions/index.d.ts"
      ],
      "ethers": [
        "./dist/ethers.d.ts"
      ],
      "viem": [
        "./dist/viem.d.ts"
      ],
      "test-utils": [
        "./dist/test-utils.d.ts"
      ]
    }
  },
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "scripts": {
    "build": "tsup"
  },
  "dependencies": {
    "@aave/env": "workspace:*",
    "@aave/graphql": "workspace:*",
    "@aave/types": "workspace:*",
    "@urql/core": "^5.2.0",
    "graphql": "^16.11.0"
  },
  "devDependencies": {
    "ethers": "^6.14.4",
    "tsup": "^8.5.0",
    "typescript": "^5.6.3",
    "viem": "^2.31.6"
  },
  "peerDependencies": {
    "ethers": "^6.14.4",
    "viem": "^2.31.6"
  },
  "peerDependenciesMeta": {
    "ethers": {
      "optional": true
    },
    "viem": {
      "optional": true
    }
  },
  "license": "MIT"
}
