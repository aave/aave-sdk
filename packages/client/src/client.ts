import type {
  HasProcessedKnownTransactionRequest,
  StandardData,
} from '@aave/graphql';
import {
  errAsync,
  invariant,
  okAsync,
  ResultAsync,
  type TxHash,
} from '@aave/types';
import {
  type AnyVariables,
  createClient,
  type Exchange,
  fetchExchange,
  type OperationResult,
  type OperationResultSource,
  type TypedDocumentNode,
  type Client as UrqlClient,
} from '@urql/core';
import { hasProcessedKnownTransaction } from './actions';
import { BatchQueryBuilder } from './batch';
import type { ClientConfig } from './config';
import { type Context, configureContext } from './context';
import { TimeoutError, UnexpectedError } from './errors';
import { Logger, LogLevel } from './logger';
import {
  isHasProcessedKnownTransactionRequest,
  type TransactionExecutionResult,
} from './types';
import { delay } from './utils';

function takeValue<T>({
  data,
  error,
}: OperationResult<StandardData<T> | undefined, AnyVariables>): T {
  invariant(data, `Expected a value, got: ${error?.message}`);
  return data.value;
}

export class AaveClient {
  /**
   * @internal
   */
  public readonly urql: UrqlClient;

  private readonly logger: Logger;

  private constructor(private readonly context: Context) {
    this.logger = Logger.named(
      this.constructor.name,
      context.debug ? LogLevel.DEBUG : LogLevel.SILENT,
    );

    this.urql = createClient({
      url: context.environment.backend,
      fetchOptions: {
        credentials: 'omit',
        headers: context.headers,
      },
      exchanges: this.exchanges(),
    });
  }

  /**
   * Create a new instance of the {@link AaveClient}.
   *
   * ```ts
   * const client = AaveClient.create({
   *   environment: production,
   *   origin: 'http://example.com',
   * });
   * ```
   *
   * @param options - The options to configure the client.
   * @returns The new instance of the client.
   */
  static create(options?: ClientConfig): AaveClient {
    return new AaveClient(configureContext(options ?? {}));
  }

  /**
   * Execute a GraphQL query operation.
   *
   * @param document - The GraphQL document to execute.
   * @param variables - The variables to pass to the operation.
   * @returns The result of the operation.
   */
  public query<TValue, TVariables extends AnyVariables>(
    document: TypedDocumentNode<StandardData<TValue>, TVariables>,
    variables: TVariables,
  ): ResultAsync<TValue, UnexpectedError> {
    const query = this.context.fragments.replaceFrom(document);
    return this.resultFrom(this.urql.query(query, variables)).map(takeValue);
  }

  /**
   * Execute a GraphQL mutation operation.
   *
   * @param document - The GraphQL document to execute.
   * @param variables - The variables to pass to the operation.
   * @returns The result of the operation.
   */
  public mutation<TValue, TVariables extends AnyVariables>(
    document: TypedDocumentNode<StandardData<TValue>, TVariables>,
    variables: TVariables,
  ): ResultAsync<TValue, UnexpectedError> {
    return this.resultFrom(this.urql.mutation(document, variables)).map(
      takeValue,
    );
  }

  // TODO update example below
  /**
   * Execute a batch of GraphQL query operations.
   *
   * @alpha This is an alpha API and may be subject to breaking changes.
   *
   * ```ts
   * const result = await client.batch((c) => [
   *   fetchAccount(c, { address: evmAddress('0x1234…') }).map(nonNullable),
   *   fetchBalancesBulk(c, {
   *     includeNative: true,
   *     tokens: [
   *       evmAddress("0x5678…"),
   *       evmAddress("0x9012…"),
   *     ],
   *   }),
   * ]);
   *
   * // const result: Result<
   * //   [
   * //     Account,
   * //     AnyAccountBalance[],
   * //   ],
   * //   UnauthenticatedError | UnexpectedError
   * // >
   * ```
   *
   * @param cb - The callback with the scoped client to execute the actions with.
   * @returns The results of all queries in the same order as they were added.
   */
  batch<T1, T2, E1 extends Error, E2 extends Error>(
    cb: (client: this) => [ResultAsync<T1, E1>, ResultAsync<T2, E2>],
  ): ResultAsync<[T1, T2], E1 | E2>;
  batch<T1, T2, T3, E1 extends Error, E2 extends Error, E3 extends Error>(
    cb: (
      client: this,
    ) => [ResultAsync<T1, E1>, ResultAsync<T2, E2>, ResultAsync<T3, E3>],
  ): ResultAsync<[T1, T2, T3], E1 | E2 | E3>;
  batch<
    T1,
    T2,
    T3,
    T4,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
    ],
  ): ResultAsync<[T1, T2, T3, T4], E1 | E2 | E3 | E4>;
  batch<
    T1,
    T2,
    T3,
    T4,
    T5,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
    E5 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
      ResultAsync<T5, E5>,
    ],
  ): ResultAsync<[T1, T2, T3, T4, T5], E1 | E2 | E3 | E4 | E5>;
  batch<
    T1,
    T2,
    T3,
    T4,
    T5,
    T6,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
    E5 extends Error,
    E6 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
      ResultAsync<T5, E5>,
      ResultAsync<T6, E6>,
    ],
  ): ResultAsync<[T1, T2, T3, T4, T5, T6], E1 | E2 | E3 | E4 | E5 | E6>;
  batch<
    T1,
    T2,
    T3,
    T4,
    T5,
    T6,
    T7,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
    E5 extends Error,
    E6 extends Error,
    E7 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
      ResultAsync<T5, E5>,
      ResultAsync<T6, E6>,
      ResultAsync<T7, E7>,
    ],
  ): ResultAsync<
    [T1, T2, T3, T4, T5, T6, T7],
    E1 | E2 | E3 | E4 | E5 | E6 | E7
  >;
  batch<
    T1,
    T2,
    T3,
    T4,
    T5,
    T6,
    T7,
    T8,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
    E5 extends Error,
    E6 extends Error,
    E7 extends Error,
    E8 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
      ResultAsync<T5, E5>,
      ResultAsync<T6, E6>,
      ResultAsync<T7, E7>,
      ResultAsync<T8, E8>,
    ],
  ): ResultAsync<
    [T1, T2, T3, T4, T5, T6, T7, T8],
    E1 | E2 | E3 | E4 | E5 | E6 | E7 | E8
  >;
  batch<
    T1,
    T2,
    T3,
    T4,
    T5,
    T6,
    T7,
    T8,
    T9,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
    E5 extends Error,
    E6 extends Error,
    E7 extends Error,
    E8 extends Error,
    E9 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
      ResultAsync<T5, E5>,
      ResultAsync<T6, E6>,
      ResultAsync<T7, E7>,
      ResultAsync<T8, E8>,
      ResultAsync<T9, E9>,
    ],
  ): ResultAsync<
    [T1, T2, T3, T4, T5, T6, T7, T8, T9],
    E1 | E2 | E3 | E4 | E5 | E6 | E7 | E8 | E9
  >;
  batch<
    T1,
    T2,
    T3,
    T4,
    T5,
    T6,
    T7,
    T8,
    T9,
    T10,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
    E5 extends Error,
    E6 extends Error,
    E7 extends Error,
    E8 extends Error,
    E9 extends Error,
    E10 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
      ResultAsync<T5, E5>,
      ResultAsync<T6, E6>,
      ResultAsync<T7, E7>,
      ResultAsync<T8, E8>,
      ResultAsync<T9, E9>,
      ResultAsync<T10, E10>,
    ],
  ): ResultAsync<
    [T1, T2, T3, T4, T5, T6, T7, T8, T9, T10],
    E1 | E2 | E3 | E4 | E5 | E6 | E7 | E8 | E9 | E10
  >;
  batch<T, E extends Error>(
    cb: (client: this) => ResultAsync<T, E>[],
  ): ResultAsync<T[], E>;
  batch(
    cb: (client: this) => ResultAsync<unknown[], unknown>[],
  ): ResultAsync<unknown[], unknown> {
    const builder = new BatchQueryBuilder();

    const client: this = Object.create(this, {
      query: {
        value: builder.addQuery,
      },
    });

    const combined = ResultAsync.combine(cb(client));
    const [document, variables] = builder.build();

    const query = this.context.fragments.replaceFrom(document);

    return this.resultFrom(this.urql.query(query, variables))
      .andTee(({ data, error }) => {
        invariant(data, `Expected a value, got: ${error?.message}`);
        builder.resolve(data);
      })
      .andThen(() => combined);
  }

  /**
   * @internal
   */
  readonly waitForSupportedTransaction = (
    result: TransactionExecutionResult,
  ): ResultAsync<TxHash, TimeoutError | UnexpectedError> => {
    if (isHasProcessedKnownTransactionRequest(result)) {
      return this.waitForTransaction(result);
    }
    return okAsync(result);
  };

  /**
   * Given the transaction hash of an Aave protocol transaction, wait for the transaction to be
   * processed by the Aave v3 API.
   *
   * Returns a {@link TimeoutError} if the transaction is not processed within the expected timeout period.
   *
   * @param result - The transaction execution result to wait for.
   * @returns The transaction hash or a TimeoutError
   */
  readonly waitForTransaction = (
    result: TransactionExecutionResult,
  ): ResultAsync<TxHash, TimeoutError | UnexpectedError> => {
    invariant(
      isHasProcessedKnownTransactionRequest(result),
      `Received a transaction result for an untracked operation. Make sure you're following the instructions in the docs.`,
    );

    return ResultAsync.fromPromise(
      this.pollTransactionStatus(result),
      (err) => {
        if (err instanceof TimeoutError || err instanceof UnexpectedError) {
          return err;
        }
        return UnexpectedError.from(err);
      },
    );
  };

  protected async pollTransactionStatus(
    request: HasProcessedKnownTransactionRequest,
  ): Promise<TxHash> {
    const startedAt = Date.now();

    while (Date.now() - startedAt < this.context.environment.indexingTimeout) {
      const processed = await hasProcessedKnownTransaction(this, request).match(
        (ok) => ok,
        (err) => {
          throw err;
        },
      );

      if (processed) {
        return request.txHash;
      }

      await delay(this.context.environment.pollingInterval);
    }
    throw TimeoutError.from(
      `Timeout waiting for transaction ${request.txHash} to be processed.`,
    );
  }

  protected exchanges(): Exchange[] {
    return [fetchExchange];
  }

  protected resultFrom<TData, TVariables extends AnyVariables>(
    source: OperationResultSource<OperationResult<TData, TVariables>>,
  ): ResultAsync<OperationResult<TData, TVariables>, UnexpectedError> {
    return ResultAsync.fromPromise(source.toPromise(), (err: unknown) => {
      this.logger.error(err);
      console.log(err);
      return UnexpectedError.from(err);
    }).andThen((result) => {
      if (result.error?.networkError) {
        return errAsync(UnexpectedError.from(result.error.networkError));
      }
      return okAsync(result);
    });
  }
}
