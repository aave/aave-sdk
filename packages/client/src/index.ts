export * from '@aave/env';
export * from '@aave/graphql';
export * from '@aave/types';
export * from './client';
export * from './config';
export * from './errors';
export * from './types';
