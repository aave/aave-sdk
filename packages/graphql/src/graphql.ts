import type {
  BigDecimal,
  BigIntString,
  BlockchainData,
  ChainId,
  Cursor,
  DateTime,
  EvmAddress,
  FixedBytes32,
  ID,
  Signature,
  TxHash,
  Void,
} from '@aave/types';
import { InvariantError } from '@aave/types';
import {
  type DocumentDecoration,
  initGraphQLTada,
  type TadaDocumentNode,
} from 'gql.tada';
import type { StandardData } from './common';
import type {
  OperationType,
  OrderDirection,
  PageSize,
  TimeWindow,
} from './enums';
import type { introspection } from './graphql-env';

export type { FragmentOf } from 'gql.tada';

export const graphql = initGraphQLTada<{
  disableMasking: true;
  introspection: introspection;
  scalars: {
    AlwaysTrue: true;
    BigDecimal: BigDecimal;
    BigInt: BigIntString;
    BlockchainData: BlockchainData;
    Boolean: boolean;
    ChainId: ChainId;
    Cursor: Cursor;
    DateTime: DateTime;
    EvmAddress: EvmAddress;
    FixedBytes32: FixedBytes32;
    Float: number;
    ID: ID;
    Int: number;
    OperationType: OperationType;
    OrderDirection: OrderDirection;
    PageSize: PageSize;
    Signature: Signature;
    String: string;
    TxHash: TxHash;
    Void: Void;
    TimeWindow: TimeWindow;
  };
}>();

/**
 * @internal
 */
export type RequestOf<Document> = Document extends DocumentDecoration<
  unknown,
  { request: infer Request }
>
  ? Request
  : never;

/**
 * @internal
 */
export type FragmentShape = NonNullable<Parameters<typeof graphql>[1]>[number];

type GetDocumentNode<
  In extends string = string,
  Fragments extends FragmentShape[] = FragmentShape[],
> = ReturnType<typeof graphql<In, Fragments>>;

/**
 * Used in unions to future-proof against new types being added.
 *
 * ⚠️ DO NOT MATCH ON THIS — it's here to block exhaustive checks.
 */
export type __FutureProofUnion = { __typename: symbol };

/**
 * @internal
 */
export type AnySelectionSet = object;

/**
 * @internal
 */
export type AnyVariables = Record<string, unknown>;

/**
 * @internal
 */
export type TypedSelectionSet<TTypename extends string = string> = {
  __typename: TTypename;
};

/**
 * @internal
 */
export type FragmentDocumentFor<
  TGqlNode extends AnySelectionSet,
  TTypename extends string = TGqlNode extends TypedSelectionSet<infer TTypename>
    ? TTypename
    : never,
  TFragmentName extends string = TTypename,
> = TadaDocumentNode<
  TGqlNode,
  AnyVariables,
  {
    fragment: TFragmentName;
    on: TTypename;
    masked: false;
  }
>;

export type RequestFrom<In extends string> = RequestOf<
  GetDocumentNode<In, FragmentShape[]>
>;

// biome-ignore lint/suspicious/noExplicitAny: simplifies necessary type assertions
export type StandardDocumentNode<Value = any, Request = any> = TadaDocumentNode<
  StandardData<Value>,
  { request: Request }
>;

/*
 * Asserts that the node is of a specific type in a union.
 *
 * ```ts
 * type A = { __typename: 'A', a: string };
 * type B = { __typename: 'B', b: string };
 *
 * const node: A | B = { __typename: 'A', a: 'a' };
 *
 * assertTypename(node, 'A');
 *
 * console.log(node.a); // OK
 * ```
 *
 * @param node - The node to assert the typename of
 * @param typename - The expected typename
 */
export function assertTypename<Typename extends string>(
  node: TypedSelectionSet,
  typename: Typename,
): asserts node is TypedSelectionSet<Typename> {
  if (node.__typename !== typename) {
    throw new InvariantError(
      `Expected node to have typename "${typename}", but got "${node.__typename}"`,
    );
  }
}
