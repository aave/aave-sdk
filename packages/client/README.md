# `@aave/client`

The official TypeScript library for interacting with Aave Protocol.

---


The `@aave/client` package contains the core functionality to query markets, execute transactions, and manage user positions across Aave lending markets and vaults. It provides a type-safe interface with built-in error handling and result types.


## Usage

```ts
import { AaveClient, evmAddress, chainId } from '@aave/client';
import { supply, userSupplies } from '@aave/client/actions';
import { sendWith } from '@aave/client/viem';

// Create client
const client = AaveClient.create();

// Query user positions
const positions = await userSupplies(client, {
  markets: [evmAddress('0x87870bca3f3fd6335c3f4ce8392d69350b4fa4e2')],
  user: evmAddress('0x742d35cc6e5c4ce3b69a2a8c7c8e5f7e9a0b1234'),
});

// Execute transactions
const result = await supply(client, {
  market: evmAddress('0x87870bca3f3fd6335c3f4ce8392d69350b4fa4e2'),
  amount: {
    erc20: {
      currency: evmAddress('0xa0b86a33e6441c8c5f0bb9b7e5e1f8bbf5b78b5c'),
      value: '1000'
    }
  },
  supplier: evmAddress('0x742d35cc6e5c4ce3b69a2a8c7c8e5f7e9a0b1234'),
  chainId: chainId(1),
})
  .andThen(sendWith(wallet))
  .andThen(client.waitForTransaction);
```
