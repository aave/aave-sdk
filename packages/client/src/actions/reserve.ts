import { type Reserve, ReserveQuery, type ReserveRequest } from '@aave/graphql';
import type { ResultAsync } from '@aave/types';
import type { AaveClient } from '../client';
import type { UnexpectedError } from '../errors';

/**
 * Fetches a specific reserve by market address, token address, and chain ID.
 *
 * ```ts
 * const result = await reserve(client, {
 *   market: evmAddress('0x87870bca3f3fd6335c3f4ce8392d69350b4fa4e2'),
 *   token: evmAddress('0xa0b86a33e6441c8c5f0bb9b7e5e1f8bbf5b78b5c'),
 *   chainId: chainId(1)
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The reserve request parameters.
 * @returns The reserve data, or null if not found.
 */
export function reserve(
  client: AaveClient,
  request: ReserveRequest,
): ResultAsync<Reserve | null, UnexpectedError> {
  return client.query(ReserveQuery, { request });
}
