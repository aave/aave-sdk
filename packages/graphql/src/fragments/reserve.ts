import type { FragmentOf } from 'gql.tada';
import { graphql } from '../graphql';
import { ChainFragment } from './chain';
import {
  CurrencyFragment,
  DecimalValueFragment,
  NativeCurrencyFragment,
  TokenAmountFragment,
} from './common';

export const EmodeInfoFragment = graphql(
  `fragment EmodeInfo on EmodeInfo {
    __typename
    maxLTV {
      ...DecimalValue
    }
    liquidationThreshold {
      ...DecimalValue
    }
    liquidationPenalty {
      ...DecimalValue
    }
  }`,
  [DecimalValueFragment],
);
export type EmodeInfo = FragmentOf<typeof EmodeInfoFragment>;

export const ReserveSupplyInfoFragment = graphql(
  `fragment ReserveSupplyInfo on ReserveSupplyInfo {
    __typename
    apy {
      ...DecimalValue
    }
    total {
      ...DecimalValue
    }
    maxLTV {
      ...DecimalValue
    }
    liquidationThreshold {
      ...DecimalValue
    }
    liquidationPenalty {
      ...DecimalValue
    }
    canBeCollateral
  }`,
  [DecimalValueFragment],
);
export type ReserveSupplyInfo = FragmentOf<typeof ReserveSupplyInfoFragment>;

export const ReserveBorrowInfoFragment = graphql(
  `fragment ReserveBorrowInfo on ReserveBorrowInfo {
    __typename
    apy {
      ...DecimalValue
    }
    total {
      ...DecimalValue
    }
    borrowCap {
      ...DecimalValue
    }
    reserveFactor
  }`,
  [DecimalValueFragment],
);
export type ReserveBorrowInfo = FragmentOf<typeof ReserveBorrowInfoFragment>;

export const ReserveUserAvailabilityFragment = graphql(
  `fragment ReserveUserAvailability on ReserveUserAvailability {
    __typename
    balance {
      ...DecimalValue
    }
    supply {
      ...DecimalValue
    }
    borrow {
      ...DecimalValue
    }
  }`,
  [DecimalValueFragment],
);
export type ReserveUserAvailability = FragmentOf<
  typeof ReserveUserAvailabilityFragment
>;

export const ReserveFragment = graphql(
  `fragment Reserve on Reserve {
    __typename
    market {
      name
      chain {
        ...Chain
      }
    }
    address
    underlyingToken {
      ...Currency
    }
    aToken {
      ...Currency
    }
    vToken {
      ...Currency
    }
    acceptsNative {
      ...NativeCurrency
    }
    size {
      ...TokenAmount
    }
    utilizationRate {
      ...DecimalValue
    }
    available {
      ...TokenAmount
    }
    usdExchangeRate
    usdOracleAddress
    supplyInfo {
      ...ReserveSupplyInfo
    }
    borrowInfo {
      ...ReserveBorrowInfo
    }
    eModeInfo {
      ...EmodeInfo
    }
    userAvailability(address: $userAddress) @include(if: $includeUserFields) {
      ...ReserveUserAvailability
    }
  }`,
  [
    ChainFragment,
    CurrencyFragment,
    NativeCurrencyFragment,
    TokenAmountFragment,
    DecimalValueFragment,
    ReserveSupplyInfoFragment,
    ReserveBorrowInfoFragment,
    EmodeInfoFragment,
    ReserveUserAvailabilityFragment,
  ],
);
export type Reserve = FragmentOf<typeof ReserveFragment>;
