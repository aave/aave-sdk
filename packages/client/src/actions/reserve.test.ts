import { assertOk, chainId, evmAddress } from '@aave/client';
import { TimeWindow } from '@aave/graphql';
import { describe, expect, it } from 'vitest';
import {
  client,
  createNewWallet,
  ETHEREUM_FORK_ID,
  ETHEREUM_MARKET_ADDRESS,
  ETHEREUM_WETH_ADDRESS,
} from '../test-utils';
import { borrowAPYHistory, reserve, supplyAPYHistory } from './reserve';

function windowToDate(window: TimeWindow): Date {
  switch (window) {
    case TimeWindow.LastDay:
      return new Date(Date.now() - 1000 * 60 * 60 * 24 - 1000 * 60 * 10);
    case TimeWindow.LastWeek:
      return new Date(Date.now() - 1000 * 60 * 60 * 24 * 8);
    case TimeWindow.LastMonth:
      return new Date(Date.now() - 1000 * 60 * 60 * 24 * 31);
    case TimeWindow.LastSixMonths:
      return new Date(Date.now() - 1000 * 60 * 60 * 24 * 183);
    case TimeWindow.LastYear:
      return new Date(Date.now() - 1000 * 60 * 60 * 24 * 366);
    default:
      throw new Error(`Unknown window: ${window}`);
  }
}

describe('Given an Aave Market reserve', () => {
  const wallet = createNewWallet();
  const windowEnum = Object.values(TimeWindow);

  describe('When fetching the reserve data', () => {
    it('Then it should return the expected reserve details', async () => {
      const result = await reserve(client, {
        market: ETHEREUM_MARKET_ADDRESS,
        chainId: ETHEREUM_FORK_ID,
        underlyingToken: ETHEREUM_WETH_ADDRESS,
        user: evmAddress(wallet.account!.address),
      });
      assertOk(result);
      expect(result.value).toMatchSnapshot({
        aToken: expect.any(Object),
        underlyingToken: {
          address: ETHEREUM_WETH_ADDRESS,
        },
        vToken: expect.any(Object),
        supplyInfo: expect.any(Object),
        borrowInfo: expect.any(Object),
        eModeInfo: expect.any(Array),
        market: expect.any(Object),
        acceptsNative: expect.any(Object),
        size: expect.any(Object),
        userState: expect.any(Object),
      });
    });
  });

  describe('When fetching the borrow APY history for it', () => {
    it.each(windowEnum)(
      'Then it should return a time series for the specified window %s',
      async (window) => {
        const result = await borrowAPYHistory(client, {
          market: ETHEREUM_MARKET_ADDRESS,
          chainId: chainId(1),
          underlyingToken: ETHEREUM_WETH_ADDRESS,
          window,
        });
        assertOk(result);
        expect(result.value?.length).toBeGreaterThan(0);
        for (const point of result.value!) {
          expect(point).toEqual(
            expect.objectContaining({
              date: expect.toBeBetweenDates(windowToDate(window), new Date()),
            }),
          );
        }
      },
    );
  });

  describe('When fetching the supply APY history for it', () => {
    it.each(windowEnum)(
      'Then it should return a time series for the specified window %s',
      async (window) => {
        const result = await supplyAPYHistory(client, {
          market: ETHEREUM_MARKET_ADDRESS,
          chainId: chainId(1),
          underlyingToken: ETHEREUM_WETH_ADDRESS,
          window,
        });
        assertOk(result);
        expect(result.value?.length).toBeGreaterThan(0);
        expect(result.value).toEqual(
          result.value?.map(() =>
            expect.objectContaining({
              date: expect.toBeBetweenDates(windowToDate(window), new Date()),
            }),
          ),
        );
      },
    );
  });
});
