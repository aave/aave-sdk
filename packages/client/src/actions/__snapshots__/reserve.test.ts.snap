// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given an Aave Market reserve > When fetching the borrow APY history for it > Then it should return a time series for the specified window LAST_WEEK 1`] = `
{
  "__typename": "Reserve",
  "aToken": Any<Object>,
  "acceptsNative": Any<Object>,
  "borrowInfo": Any<Object>,
  "eModeInfo": Any<Array>,
  "flashLoanEnabled": true,
  "isFrozen": false,
  "isPaused": false,
  "isolationModeConfig": null,
  "market": Any<Object>,
  "size": Any<Object>,
  "supplyInfo": Any<Object>,
  "underlyingToken": {
    "__typename": "Currency",
    "address": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
    "chainId": 99999999,
    "decimals": 18,
    "imageUrl": "https://token-logos.family.co/asset?id=99999999:0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2&token=WETH",
    "name": "Wrapped Ether",
    "symbol": "WETH",
  },
  "usdExchangeRate": "3682.78",
  "usdOracleAddress": "0x5424384B256154046E9667dDFaaa5e550145215e",
  "userState": Any<Object>,
  "vToken": Any<Object>,
}
`;

exports[`Given an Aave Market reserve > When fetching the reserve data > Then it should return the expected reserve details 1`] = `
{
  "__typename": "Reserve",
  "aToken": Any<Object>,
  "acceptsNative": Any<Object>,
  "borrowInfo": Any<Object>,
  "eModeInfo": Any<Array>,
  "flashLoanEnabled": true,
  "isFrozen": false,
  "isPaused": false,
  "isolationModeConfig": null,
  "market": Any<Object>,
  "size": Any<Object>,
  "supplyInfo": Any<Object>,
  "underlyingToken": {
    "__typename": "Currency",
    "address": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
    "chainId": 99999999,
    "decimals": 18,
    "imageUrl": "https://token-logos.family.co/asset?id=99999999:0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2&token=WETH",
    "name": "Wrapped Ether",
    "symbol": "WETH",
  },
  "usdExchangeRate": "3682.78",
  "usdOracleAddress": "0x5424384B256154046E9667dDFaaa5e550145215e",
  "userState": Any<Object>,
  "vToken": Any<Object>,
}
`;
